### PR TITLE
Add and implement a couple of fallback defines for MSVC - What was missed

### PR DESCRIPTION
**Pull request by Orvid**
When #264 was merged, some confusion appears to have occurred over whether the defines were needed for the final release of VS 2015. They are needed, so this adds them.

This also adds the change that was excluded when #297 was merged.

### CODE REVIEW

File: folly/AtomicLinkedList.h [hunk 1]
```diff
@@ -67,7 +68,11 @@ class AtomicLinkedList {
   }
 
   bool empty() const {
+#ifdef MSVC_NO_NONVOID_ATOMIC_IF
+    return head_.load() == nullptr;
+#else
     return head_ == nullptr;
+#endif
```
Comment (JoelMarcey) [inline folly/AtomicLinkedList.h]: If we're supporting broken compilers, just call .load() explicitly for all compilers. And add a comment explaining why.

File: folly/Format-inl.h [hunk 2]
```diff
@@ -774,9 +774,7 @@ class TryFormatValue {
 
 template <class T>
 class TryFormatValue<
-  T,
-  typename std::enable_if<
-      0 < sizeof(FormatValue<typename std::decay<T>::type>)>::type>
+  T, typename FormatValue<typename std::decay<T>::type>::type>
```
Comment (JoelMarcey) [inline folly/Format-inl.h]: Why is this necessary?

File: folly/Portability.h [hunk 3]
```diff
@@ -334,6 +334,17 @@ inline size_t malloc_usable_size(void* ptr) {
...
 
+#if defined(_MSC_VER) && _MSC_FULL_VER <= 190023026 // 2015 RTM or below
+// MSVC2015's initial release produces errors if you try
+// to do `if (std::atomic<char*>)`, while `if (std::atomic<void*>)`
+// works just fine.
+// Bug Report: https://connect.microsoft.com/VisualStudio/feedback/details/1464842
+# define MSVC_NO_NONVOID_ATOMIC_IF 1
+// 2015 RTM doesn't support in-class initialization of static constexpr members.
+// Attempting to do this produces an error informing you that this is not yet supported.
+# define MSVC_NO_STATIC_INCLASS_CONSTEXPR_INITIALIZATION 1
```
Comment (JoelMarcey) [inline folly/Portability.h]: Are there no newer MSVC2015 compilers? Are we going to support broken early versions of fixed compilers?

File: folly/Singleton.cpp [hunk 4]
```diff
@@ -22,7 +22,11 @@ namespace folly {
 
 namespace detail {
 
+#ifdef MSVC_NO_STATIC_INCLASS_CONSTEXPR_INITIALIZATION
+const std::chrono::seconds SingletonHolderBase::kDestroyWaitTime{5};
+#else
 constexpr std::chrono::seconds SingletonHolderBase::kDestroyWaitTime;
+#endif
```
Comment (JoelMarcey) [inline folly/Singleton.cpp]: Are there no newer MSVC2015 compilers? Are we going to support broken early versions of fixed compilers?

File: folly/Singleton.h [hunk 5]
```diff
@@ -200,7 +200,11 @@ class SingletonHolderBase {
   virtual void destroyInstance() = 0;
 
  protected:
+#ifdef MSVC_NO_STATIC_INCLASS_CONSTEXPR_INITIALIZATION
+  static const std::chrono::seconds kDestroyWaitTime;
+#else
   static constexpr std::chrono::seconds kDestroyWaitTime{5};
+#endif
```
Comment (JoelMarcey) [inline folly/Singleton.h]: If we're supporting broken compilers, just make this `const` rather than `constexpr` for all compilers. And add a comment explaining why.

Or keep this as `constexpr` but move it out of the class and into the namespace, renaming it to

`kSingletonHolderBaseDestroyWaitTime`

And add a comment why it can't go into the class.

File: folly/Portability.h [hunk 6]
```diff
@@ -334,6 +334,17 @@ inline size_t malloc_usable_size(void* ptr) {
...
 #endif
 
+#if defined(_MSC_VER) && _MSC_FULL_VER <= 190023026 // 2015 RTM or below
+// MSVC2015's initial release produces errors if you try
+// to do `if (std::atomic<char*>)`, while `if (std::atomic<void*>)`
+// works just fine.
+// Bug Report: https://connect.microsoft.com/VisualStudio/feedback/details/1464842
+# define MSVC_NO_NONVOID_ATOMIC_IF 1
+// 2015 RTM doesn't support in-class initialization of static constexpr members.
+// Attempting to do this produces an error informing you that this is not yet supported.
```
Comment (JoelMarcey) [inline folly/Portability.h]: Lint - make this line less than 80 chars

**Comment by JoelMarcey** [comment]
@Orvid I add some comments from a reviewer internally.

File: folly/Portability.h (hunk 3 above)
Comment (Orvid) [inline folly/Portability.h]: There are no newer versions currently.

Comment (Orvid) [inline folly/Portability.h]: As soon as a newer version is released that fixes these, I fully intend to remove the workarounds.

File: folly/Format-inl.h (hunk 2 above)
Comment (Orvid) [inline folly/Format-inl.h]: I'm not certain,  but I believe this is due to MSVC not yet supporting expression SFINAE. Whatever the actual reason MSVC doesn't like the sizeof on an invalid type.

**Comment by Orvid** [comment]
Moved internally to D2901301.
